File: Challenge-005-Queue/python/main.py
#!python3
class Queue:
    def __init__(self):
        self.items = []

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if self.is_empty():
            print("Queue underflow")
            return None
        return self.items.pop(0)

    def peek(self):
        if self.is_empty():
            print("Queue is empty")
            return None
        return self.items[0]

    def is_empty(self):
        return len(self.items) == 0

# Ejemplo de uso
if __name__ == "__main__":
    queue = Queue()
    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(30)

    print("Frente:", queue.peek())      # 10
    print("Dequeue:", queue.dequeue())  # 10
    print("Nuevo frente:", queue.peek()) # 20

File: Challenge-005-Queue/rust/src/main.rs
/* Template for Rust program */
struct Queue<T> {
    items: Vec<T>,
}

impl<T> Queue<T> {
    fn new() -> Self {
        Queue { items: Vec::new() }
    }

    fn enqueue(&mut self, item: T) {
        self.items.push(item);
    }

    fn dequeue(&mut self) -> Option<T> {
        if self.is_empty() {
            println!("Queue underflow");
            None
        } else {
            Some(self.items.remove(0))
        }
    }

    fn peek(&self) -> Option<&T> {
        if self.is_empty() {
            println!("Queue is empty");
            None
        } else {
            self.items.first()
        }
    }

    fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

fn main() {
    let mut queue = Queue::new();
    queue.enqueue(10);
    queue.enqueue(20);
    queue.enqueue(30);

    println!("Frente: {:?}", queue.peek());      // Some(10)
    println!("Dequeue: {:?}", queue.dequeue());  // Some(10)
    println!("Nuevo frente: {:?}", queue.peek()); // Some(20)
}